Extracted actor Voices from Selectable into own Voiced trait

reaperrr opened the pull request:
And implemented `IVoiced` interface to move `Voiced` to Common.

reaperrr committed:
Extract actor voice set into Voiced trait

abcdefg30 commented:
> • track down and fix the issue that Select voice is played twice when actor(s) are selected via group key.

Would fix #5770 then.

Mailaender commented:
How about calling it `Voiced`?

abcdefg30 commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -44,8 +44,8 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
-			var selectable = a.Info.Traits.GetOrDefault<SelectableInfo>();
-			return selectable != null && selectable.Voice != null;
+			var hasVoices = a.HasTrait<IWithVoices>();
+			return hasVoices && a.TraitOrDefault<IWithVoices>().VoiceSet != null;
```
What about something like
```c#
var voice = a.TraitOrDefault<IWithVoices>();
return voice != null && voice.VoiceSet != null;

Mailaender replied:
Saves one trait lookup.

Phrohdoh replied:
`TraitOrDefault<T>` doesn't make sense if the actor `HasTrait<T>`.

penev92 replied:
What @abcdefg30 said.

abcdefg30 commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -56,9 +56,11 @@ public static bool HasVoice(this Actor a, string voice)
+				return null;
+
+			var v = a.TraitOrDefault<IWithVoices>().VoiceSet;
 			return (v == null) ? null : a.World.Map.Rules.Voices[v.ToLowerInvariant()];
```
Same question here.

Phrohdoh commented:
AFAIK With* is typically reserved for rendering traits.

Mailaender commented:
I think you can also deprecate the weird

```yaml
Selectable:
    Selectable: false
```

setting then.

reaperrr committed:
Upgrade rules for moving Voice to Voiced trait

reaperrr commented:
@Mailaender: That setting is weird indeed, but I want to deal with Selectable in a later, dedicated PR anyway (separating selection bounds from visible selection box), so I don't want to touch that in this PR.

pchote commented:
IIRC that Selectable property was there as a workaround for #2922, which has now been fixed.

abcdefg30 commented:
Needs a rebase.

penev92 commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -56,9 +56,7 @@ public static bool HasVoice(this Actor a, string voice)
-			var selectable = a.Info.Traits.GetOrDefault<SelectableInfo>();
-			if (selectable == null) return null;
-			var v = selectable.Voice;
+			var v = a.TraitOrDefault<IVoiced>().VoiceSet;
```
You need to keep the null check here because `TraitOrDefault` can give you null, and you have no guarantees that whoever calls this method will check beforehand.

penev92 commented on mods/ts/rules/civilian-infantry.yaml:
```diff
@@ -284,8 +293,8 @@ DOGGIE:
-	Selectable:
-		Voice: Fiend
+	Voiced:
+		VoiceSet: Fiend
```
Why is this no longer selectable?

penev92 commented on mods/ts/rules/civilian-infantry.yaml:
```diff
@@ -314,8 +323,8 @@ VISSML:
-	Selectable:
-		Voice: Fiend
+	Voiced:
+		VoiceSet: Fiend
```
And this?

penev92 commented:
Changes look OK. Will test ingame when I can.

reaperrr commented:
Updated (reimplemented null check in WorldUtils.GetVoices) and rebased.

pchote commented on OpenRA.Game/Sound/Sound.cs:
```diff
@@ -364,30 +364,30 @@ public static float VideoSeekPosition
 		}
 
-		public static bool PlayVoice(string phrase, Actor voicedUnit, string variant)
+		public static bool PlayVoice(string phrase, Actor voicedActor, string variant)
```
It would be great if we could move this into the IVoiced trait implementations themselves.  The sound code really shouldn't be digging around inside actors.

reaperrr replied:
Done.

reaperrr committed:
Move PlayVoice and PlayVoiceLocal from Sound to Voiced.

Move HasVoice, HasVoices and GetVoices from WorldUtils to Voiced.

reaperrr commented:
Updated.

penev92 commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -44,8 +44,8 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
 		{
-			var selectable = a.Info.Traits.GetOrDefault<SelectableInfo>();
-			return selectable != null && selectable.Voice != null;
+			var voice = a.TraitOrDefault<IVoiced>();
```
Let's not assume the actor has only one such trait and keep to
`var voice = a.TraitsImplementing<IVoiced>().FirstOrDefault();`

penev92 commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -56,9 +56,11 @@ public static bool HasVoice(this Actor a, string voice)
-			var selectable = a.Info.Traits.GetOrDefault<SelectableInfo>();
-			if (selectable == null) return null;
-			var v = selectable.Voice;
+			var voiced = a.TraitOrDefault<IVoiced>();
```
Same here.

pchote commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -44,8 +44,8 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
 
 		public static bool HasVoices(this Actor a)
```
It should be relatively simple to remove these three now, but this might be scope creeping to much.

penev92 replied:
I was going to object, but I see these three methods have a total of 2 usages, so I'm, inclined to let it go.

Phrohdoh replied:
My opinion is that this is creep and should be done another time.

reaperrr replied:
> It should be relatively simple to remove these three now, but this might be scope creeping to much.

Done

reaperrr commented:
Updated.

penev92 commented on OpenRA.Game/Selection.cs:
```diff
@@ -57,9 +57,10 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
-			var voicedUnit = actors.FirstOrDefault(a => a.Owner == world.LocalPlayer && a.IsInWorld && a.HasVoices());
-			if (voicedUnit != null)
-				Sound.PlayVoice("Select", voicedUnit, voicedUnit.Owner.Country.Race);
+			var voicedActor = actors.FirstOrDefault(a => a.Owner == world.LocalPlayer && a.IsInWorld && (a.TraitsImplementing<IVoiced>() != null));
```
`.TraitsImplementing()` should never be able to return `null`. The proper check here is probably `.TraitsImplementing().Any()`.
But I suggest using `.HasTrait<IVoiced>()`, which is much simpler ;)

penev92 commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -73,10 +53,15 @@ public static void PlayVoiceForOrders(this World w, Order[] orders)
 					continue;
 
+				var hasVoice = o.Subject.TraitsImplementing<IVoiced>();
+				if (hasVoice == null)
```
Suggestion: rename `hasVoice` to `voices` here and use it instead of doing another `o.Subject.TraitsImplementing<IVoiced>()` in the `foreach` 3 lines below.

penev92 commented on OpenRA.Mods.Common/Traits/Cargo.cs:
```diff
@@ -207,7 +207,7 @@ public string VoicePhraseForOrder(Actor self, Order order)
 				return null;
 
-			return self.HasVoice("Unload") ? "Unload" : "Move";
+			return self.TraitsImplementing<IVoiced>().FirstOrDefault().HasVoice(self, "Unload") ? "Unload" : "Move";
```
Please change this to
```
return self.TraitsImplementing<IVoiced>().Any(x => x.HasVoice(self, "Unload")) ? "Unload" : "Move";
```
`.FirstOrDefault().HasVoice()` is a potential NullReferenceException.

penev92 commented on OpenRA.Mods.Common/Traits/Voiced.cs:
```diff
@@ -0,0 +1,84 @@
+			return Sound.PlayPredefined(voicedActor.World.Map.Rules, null, voicedActor, type, phrase, variant, false, pos, volume, true);
+		}
+
+		public bool HasVoices(Actor actor)
```
You can delete this method. Nothing is using it any more and there isn't a point to checking if the actor has trait X inside trait X.

penev92 commented on OpenRA.Mods.Common/Traits/Voiced.cs:
```diff
@@ -0,0 +1,84 @@
+			return v != null && v.Voices.ContainsKey(voice);
+		}
+
+		public SoundInfo GetVoices(Actor actor)
```
This method and the above can be simplified to a single
```
		public bool HasVoice(Actor self, string voice)
		{
			if (string.IsNullOrEmpty(Info.VoiceSet))
				return false;

			var voices = self.World.Map.Rules.Voices[Info.VoiceSet.ToLowerInvariant()];
			return voices != null && voices.Voices.ContainsKey(voice);
		}
```

penev92 commented on OpenRA.Mods.Common/Traits/Voiced.cs:
```diff
@@ -0,0 +1,84 @@
+
+		public Voiced(Actor self, VoicedInfo info)
+		{
+			Info = info;
```
You can get rid of the actors passed as arguments to all the methods if you cache `self` here and use that.

pchote replied:
Convention is to have the first argument to public methods be `Actor self`.

penev92 commented on OpenRA.Game/Traits/TraitsInterfaces.cs:
```diff
@@ -109,6 +109,15 @@ public interface IDisableMove { bool MoveDisabled(Actor self); }
+		bool PlayVoice(string phrase, Actor voicedActor, string variant);
+		bool PlayVoiceLocal(string phrase, Actor voicedActor, string variant, WPos pos, float volume);
+		bool HasVoices(Actor actor);
+		bool HasVoice(Actor actor, string voice);
```
Once you address my comments on `Voiced` this whould only have `bool HasVoice(string voice);`

Phrohdoh replied:
Per @pchote's comment; keep this as-is.

penev92 replied:
Actually, still remove `HasVoices()` and chenge this to `Actor self`.

penev92 commented on OpenRA.Game/Traits/TraitsInterfaces.cs:
```diff
@@ -109,6 +109,15 @@ public interface IDisableMove { bool MoveDisabled(Actor self); }
+		bool PlayVoiceLocal(string phrase, Actor voicedActor, string variant, WPos pos, float volume);
+		bool HasVoices(Actor actor);
+		bool HasVoice(Actor actor, string voice);
+		string VoiceSet { get; }
```
Please move this property before the method declarations (conventions state properties go before methods).

pchote commented on OpenRA.Mods.Common/Traits/Voiced.cs:
```diff
@@ -0,0 +1,84 @@
+			return Sound.PlayPredefined(voicedActor.World.Map.Rules, null, voicedActor, type, phrase, variant, true, WPos.Zero, 1f, true);
+		}
+
+		public bool PlayVoiceLocal(string phrase, Actor voicedActor, string variant, WPos pos, float volume)
```
* Is there any reason why `pos` wouldn't be the actor position?
* Please rename `voicedActor` to `self` and move to the first arg.
* `self` will not be null, and `mi == this`, so you can make some cleanups in the method body.

reaperrr replied:
> Is there any reason why pos wouldn't be the actor position?

Not really, that was just copy-pasta.

pchote commented on OpenRA.Mods.Common/Traits/Voiced.cs:
```diff
@@ -0,0 +1,84 @@
+			if (voicedActor == null || phrase == null)
+				return false;
+
+			var mi = voicedActor.TraitOrDefault<IVoiced>();
```
Likewise here, `mi == this` and `voicedActor` &rarr; `self`.

penev92 commented on OpenRA.Mods.Common/Traits/Voiced.cs:
```diff
@@ -0,0 +1,84 @@
+			if (mi == null || mi.VoiceSet == null)
+				return false;
+
+			var type = mi.VoiceSet.ToLowerInvariant();
```
You can replace the last 4 lines with
```
if (string.IsNullOrEmpty(Info.VoiceSet))
    return false;
var type = Info.VoiceSet.ToLowerInvariant();
```

reaperrr committed:
Cleaned up IVoiced-related code.

Added Volume control and descriptions to Voiced.

Streamline voice checks in WorldUtils and DeathSounds.

reaperrr commented:
Updated.
Adressed everything that was pointed out, I think.

penev92 commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -70,13 +50,19 @@ public static void PlayVoiceForOrders(this World w, Order[] orders)
+					continue;
+
+				var voices = orderSubject.TraitsImplementing<IVoiced>();
+				if (voices == null)
```
`if (!voices.Any())`

penev92 commented on OpenRA.Mods.Common/Traits/Sound/DeathSounds.cs:
```diff
@@ -40,11 +40,11 @@ public void Killed(Actor self, AttackInfo e)
-			if (info.DeathTypes.Contains(e.Warhead.DeathType) ||
-				(!info.DeathTypes.Any() && !self.Info.Traits.WithInterface<DeathSoundsInfo>().Any(dsi => dsi.DeathTypes.Contains(e.Warhead.DeathType))))
-				Sound.PlayVoiceLocal(info.DeathSound, self, self.Owner.Country.Race, cp, info.VolumeMultiplier);
+			foreach (var voiced in self.TraitsImplementing<IVoiced>())
```
Please move the `foreach` inside the `if`.

penev92 commented:
Code looks good and everything sounds good in all 4 mods.
:+1: 
Also a very nice cleanup.

reaperrr commented:
Should I squash some (or all) of the commits before this is merged?

abcdefg30 commented:
Yeah, I think squashing some of the commits together is a good idea, but you don't need to.

reaperrr commented:
Squashed some of the commits and (finally) added a fix that closes #8121.

pchote commented on OpenRA.Mods.Common/Traits/Cargo.cs:
```diff
@@ -207,7 +207,7 @@ public string VoicePhraseForOrder(Actor self, Order order)
 				return null;
 
-			return self.HasVoice("Unload") ? "Unload" : "Move";
+			return self.TraitsImplementing<IVoiced>().Any(x => x.HasVoice(self, "Unload")) ? "Unload" : "Move";
```
This is the only trait that hardcodes a fallback to "Move".  How about we fix the voice sets and remove this?

pchote commented:
Having the `foreach (var voiced in self.TraitsImplementing<IVoiced>())` everywhere feels a bit ugly.  Could you add a `PlayVoice` (and maybe also `HasVoice`?) extension method to ActorExts, and then use `actor.PlayVoice` instead?

reaperrr committed:
Adds PlayVoice, PlayVoiceLocal and HasVoice to ActorExts.

Makes Cargo unload voice customisable.

reaperrr committed:
Fix Unload definition in voice sets.

reaperrr commented:
Updated.

reaperrr committed:
Moved Voice-related extensions to VoiceExts.

Note: This is a work-around until Selectable can be moved to Mods.Common, which is when the voice extensions should be moved back to ActorExts.

Pulled phrase check before foreach in PlayVoice ActorExts.
Removed superflous actor parameter from PlayVoice/PlayVoiceLocal.

Simplified PlayVoice extensions.

variant is no longer customisable, as all current usages use self.Owner.Country.Race anyway.

reaperrr commented:
Updated.

abcdefg30 commented:
Sorry, but this needs a rebase again.

pchote commented on OpenRA.Mods.Common/ActorExts.cs:
```diff
@@ -80,6 +80,39 @@ public static Target ResolveFrozenActorOrder(this Actor self, Order order, Color
 			return Target.Invalid;
 		}
 
+		public static void PlayVoice(this Actor self, Actor actor, string phrase, string variant)
```
The `actor` parameters here and below aren't used (and don't make a lot of sense?).  Could you please remove them?

reaperrr replied:
I misunderstood something, will fix this.

pchote commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -70,13 +50,15 @@ public static void PlayVoiceForOrders(this World w, Order[] orders)
-					if (Sound.PlayVoice(v.VoicePhraseForOrder(o.Subject, o),
-						o.Subject, o.Subject.Owner.Country.Race))
-						return;
+				foreach (var voice in orderSubject.TraitsImplementing<IVoiced>())
```
This should be able to use the new extension methods.

reaperrr replied:
This and Selection are still in Game, while ActorExts are in Common. I could move the PlayVoice/HasVoice to Game of course (though I don't know into which file), but I believe the long-term goal was rather to move more stuff from the Game engine to Common, right?

pchote commented on OpenRA.Game/Selection.cs:
```diff
@@ -57,9 +57,10 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
-				Sound.PlayVoice("Select", voicedUnit, voicedUnit.Owner.Country.Race);
+			var voicedActor = actors.FirstOrDefault(a => a.Owner == world.LocalPlayer && a.IsInWorld && a.HasTrait<IVoiced>());
+			if (voicedActor != null)
+				foreach (var voice in voicedActor.TraitsImplementing<IVoiced>())
```
This should be able to use the new extension methods.

pchote commented:
Looks like we are converging to a nice clean up here :)  I've pointed out a couple of places you have missed, which should be easy fixups.

reaperrr commented:
Updated and rebased.

Someone notify me when this is otherwise ready to merge, then I'll squash some commits (unless you think it's fine the way it is).

pchote commented on OpenRA.Game/VoiceExts.cs:
```diff
@@ -0,0 +1,74 @@
+{
+	public static class VoiceExts
+	{
+		public static void PlayVoice(this Actor self, string phrase, string variant)
```
AFAIK the variant here and below are always just the owner's race.  We can probably drop that argument and simplify this a step further.

pchote replied:
We probably want to keep the variant string on the lower level code for now though, as they are used for different things by e.g. notifications.

obrakmann commented on OpenRA.Mods.Common/Traits/Sound/AnnounceOnBuild.cs:
```diff
@@ -19,7 +19,7 @@ public class AnnounceOnBuild : INotifyBuildComplete
 		public void BuildingComplete(Actor self)
 		{
-			Sound.PlayVoice("Build", self, self.Owner.Country.Race);
+			self.PlayVoice("Build", self.Owner.Country.Race);
```
This and the "Kill" below should probably also be moved into the Info class, like the "Unload" above. Doesn't have to be in this PR, though, just sayin'.

obrakmann commented:
All orders except normal move orders don't generate a response from the receiver of the order. AttackMoves, capturing, entering transports are all silent, both in td and ra, so I assume everywhere.

reaperrr committed:
Made AnnounceOnKill and AnnounceOnBuild voices customisable.

reaperrr commented:
Adressed everything that was mentioned.
I had to revert the changes to VoicePhraseForOrder though, because they were responsible for the issue @obrakmann pointed out

penev92 commented:
I've been keeping track of the new changes (thanks for making them in new commits), so my +1 still stands.

Mailaender commented on OpenRA.Game/Traits/Selectable.cs:
```diff
@@ -20,7 +20,6 @@ public class SelectableInfo : ITraitInfo
 		public readonly bool Selectable = true;
```
You can remove this now.

Mailaender replied:
However this pull request is already large enough.

Mailaender commented:
:+1: Thanks. This is a valueable cleanup. Works as promised. :white_check_mark:

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d8cbdb9243e1b61414db4344106be04f8e188a05...d49882200e5762d26480fc40cd7bcae72d165f48)
